core/core.go

```diff
@@ -49,22 +50,32 @@ func NewCore(db ethdb.Database, config *Config, isLocalBlock func(block *types.H
 		quit:   make(chan struct{}),
 	}
 
-	futureHeaders, _ := types.NewTimedCache(maxFutureHeaders, futureHeaderTtl)
+	futureHeaders, _ := timedcache.NewWithEvict(maxFutureHeaders, futureHeaderTtl, c.onFhCacheEvict)
 	c.futureHeaders = futureHeaders
 
 	go c.updateFutureHeaders()
 	return c, nil
 }
 
+func (c *Core) onFhCacheEvict(key, value interface{}) {
+	if c.futureHeaders.Len() == 0 {
+		// If the future header cache has drained, signal
+		// to the downloader to resume
+		c.sl.downloaderWaitFeed.Send(false)
+	}
+}
+
 func (c *Core) InsertChain(blocks types.Blocks) (int, error) {
 	nodeCtx := common.NodeLocation.Context()
 	domWait := false
 	for i, block := range blocks {
-		isCoincident := c.sl.engine.IsDomCoincident(block.Header())
 		// Write the block body to the CandidateBody database.
 		rawdb.WriteCandidateBody(c.sl.sliceDb, block.Hash(), block.Body())
-
-		if !isCoincident && !domWait {
+		// if this (or any prior) block is coincident, we need to wait for our dom to append
+		domWait = domWait && c.sl.engine.IsDomCoincident(block.Header())
```

Comment:
It seems like this evaluation on `domWait` is evaluating incorrectly. Wouldn't we prefer `||`?

Reply:
Good catch.. yes that was supposed to be `||`